flink/v1.14/flink/src/test/java/org/apache/iceberg/flink/TestFlinkUpsert.java

```diff
@@ -0,0 +1,302 @@
+          parameters.add(new Object[] {catalogName, baseNamespace, format, isStreaming});
+        }
+      }
+    }
+    return parameters;
+  }
+
+  @Override
+  protected TableEnvironment getTableEnv() {
+    if (tEnv == null) {
+      synchronized (this) {
+        EnvironmentSettings.Builder settingsBuilder = EnvironmentSettings
+            .newInstance();
+        if (isStreamingJob) {
+          settingsBuilder.inStreamingMode();
+          StreamExecutionEnvironment env = StreamExecutionEnvironment
+              .getExecutionEnvironment(MiniClusterResource.DISABLE_CLASSLOADER_CHECK_CONFIG);
+          env.enableCheckpointing(400);
+          env.setMaxParallelism(2);
+          env.setParallelism(2);
+          tEnv = StreamTableEnvironment.create(env, settingsBuilder.build());
+        } else {
+          settingsBuilder.inBatchMode();
+          tEnv = TableEnvironment.create(settingsBuilder.build());
+        }
+      }
+    }
+    return tEnv;
+  }
+
+  @Override
+  @Before
+  public void before() {
+    super.before();
+    sql("CREATE DATABASE IF NOT EXISTS %s", flinkDatabase);
+    sql("USE CATALOG %s", catalogName);
+    sql("USE %s", DATABASE);
+  }
+
+  @Override
+  @After
+  public void clean() {
+    sql("DROP DATABASE IF EXISTS %s", flinkDatabase);
+    BoundedTableFactory.clearDataSets();
+    super.clean();
+  }
+
+  @Test
+  public void testUpsertAndQuery() {
+    String tableName = "test_upsert_query";
+
+    sql("CREATE TABLE %s(id INT NOT NULL, province STRING NOT NULL, dt DATE, PRIMARY KEY(id,province) NOT ENFORCED) " +
+            "PARTITIONED BY (province) WITH %s",
+        tableName, toWithClause(tableUpsertProps));
+
+    try {
+      sql("INSERT INTO %s VALUES " +
+          "(1, 'a', TO_DATE('2022-03-01'))," +
+          "(2, 'b', TO_DATE('2022-03-01'))," +
+          "(1, 'b', TO_DATE('2022-03-01'))",
+          tableName);
+
+      sql("INSERT INTO %s VALUES " +
+          "(4, 'a', TO_DATE('2022-03-02'))," +
+          "(5, 'b', TO_DATE('2022-03-02'))," +
+          "(1, 'b', TO_DATE('2022-03-02'))",
+          tableName);
+
+      List<Row> result = sql("SELECT * FROM %s WHERE dt < '2022-03-02'", tableName);
+
+      Assert.assertEquals("result should have 2 rows!", 2, result.size());
+
+      result = sql("SELECT * FROM %s WHERE dt < '2022-03-03'", tableName);
+
+      Assert.assertEquals("result should have 5 rows!", 5, result.size());
+
+      sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    } finally {
+      sql("DROP TABLE IF EXISTS %s.%s", flinkDatabase, tableName);
+    }
+  }
+
+  @Test
+  public void testPrimaryKeyKEqualToPartitionKey() {
+    // This is an SQL based reproduction of TestFlinkIcebergSinkV2#testUpsertOnDataKey
+    String tableName = "upsert_on_data_key";
+    try {
+      sql("CREATE TABLE %s(id INT NOT NULL, data STRING NOT NULL, PRIMARY KEY(data) NOT ENFORCED) " +
+              "PARTITIONED BY (data) WITH %s",
```

Comment:
As I discussed with @kbendick , I doubt whether it is reasonable that the `parmary key` is exactly the same as the `partition key`. In a real production environment, is whether it makes sense to do so.

Reply:
If this shouldn't exist, should we throw an exception in this case, of course, not here in Test, but in other place.

Reply:
I would hold off on doing that the more that I think about it, as we have several tests that test that behavior.

So in my opinion, to begin throwing an exception during a potential patch release (which this commit would be a part of) doesn't make sense.

But it's something we could consider for afterwards. Since it wound up not causing problems, in that we didn't have to make any specific code changes to make this work like we had thought at one point we might, I don't see a reason to disallow it given that it has been allowed.

If there were more overhead to allowing this scenario (additional code, etc), then I would agree we should remove it. But since it's already allowed, at the least we shouldn't remove it in this patch.

Reply:
if we agree that this is pathological case, can we raise an issue for the follow-up change to throw an exception for the next minor release of 0.14.0 and make sure release note is tracked.

Reply:
+1 to marking it as a follow up issue if we do anything about it.

I do agree it's strange, but until it causes us extra work, I'd rather keep existing functionality working. In particular, if a user was upserting on a table partitioned by date, to have one roll up record per date, I wouldn't know how to write an instructive error message to explain to them how to resolve that issue in a generic way.

Granted it is somewhat pathological behavior, but many people who are coming from Hive continue to use their existing Hive partitioning, which can have these abnormal cases in them. If it's not any additional work to us, I don't like the idea of breaking compatibility.

Reply:
While it is probably not a good idea to do this, I don't think it is actually incorrect. I would probably leave the test in and not worry about it.